copy/copy.go

```diff
@@ -83,10 +86,23 @@ func (d *digestingReader) Read(p []byte) (int, error) {
 type Options struct {
 	RemoveSignatures bool   // Remove any pre-existing signatures. SignBy will still add a new signature.
 	SignBy           string // If non-empty, asks for a signature to be added during the copy, and specifies a key ID, as accepted by signature.NewGPGSigningMechanism().SignDockerManifest(),
+	ReportWriter     io.Writer
 }
 
 // Image copies image from srcRef to destRef, using policyContext to validate source image admissibility.
 func Image(ctx *types.SystemContext, policyContext *signature.PolicyContext, destRef, srcRef types.ImageReference, options *Options) error {
+	var report bool
+	var writeReport func(f string, a ...interface{})
+	if options.ReportWriter != nil {
+		writeReport = func(f string, a ...interface{}) {
+			fmt.Fprintf(options.ReportWriter, f, a...)
+		}
+		report = true
+	} else {
+		writeReport = func(f string, a ...interface{}) {
+			fmt.Fprintf(ioutil.Discard, f, a...)
+		}
+	}
```

Comment:
Isn’t this more or less
```go
if options.ReportWriter == nil {
    options.ReportWriter = ioutil.Discard // note that options is a struct = our local copy, not a pointer, so we can do this
}
```
and then `s/writeReport(/fmt.Fprintf(options.ReportWriter, /` ? That would also catch the missing format argument.  Even the `report` bool could go away, at some CPU cost, if we simply passed the `options.ReportWriter/ioutil.Discard/` down to `copyBlob`.

Alternatively, there must be a wayto tell code analysis tools them that `f` is a format string; and having a `writeReport` without the extra argument _is_ shorter. It would be nice to move this a bit out of line of `copy.Image`, `writeReport := prepareProgressReporter(options)` perhaps.